### ISSUE TITLE
Update docs? Intel no longer charges for MKL

### ISSUE THREAD

**Issue by kkm000** [kaldi-asr/kaldi#3078]
[At the least, for Linux](https://software.intel.com/en-us/articles/installing-intel-free-libs-and-python-apt-repo) it's freely available. I installed it used it for a while when my compiler license expired.

Should we explain this option better in the docs? They are very performant, at the least on Intel cores, and take advantage of the particular chip the run on. I can do that; at the least, I went through the process and it works.

**Comment by danpovey** [comment]
Yes, please do.  An issue that has been hanging over us for a while, that
is related, is doing a better job of making use of faster-than-ATLAS BLAS
versions, e.g. openblas or intel mkl, and steering users towards those in
the setup process.  @jtrmal was going to look at this but I'm sure he will
appreciate it if you take it from him.

**Comment by kkm000** [comment]
Sure, will do.

/cc @jtrmal

**Comment by jtrmal** [comment]
yes, thanks!
I was happy with MKL, but I thought it was free only for community (i.e.
OSS) products (or something like that). OTOH, Kaldi is OSS and no-one is
forced to use MKL, so I don't see any reason not to make it more
recommended.
y.

**Comment by jtrmal** [comment]
And of course, my knowledge might be obsolete. I didn;t check.
y.

**Comment by kkm000** [comment]
My understanding it's available for free to everyone, with the whole Performance Libraries line. Only Studio XE (the compiler, profilers etc) are licensed for OSS via a special program.

Question: when installing MKL, should I assume 64-bit architecture? I want to write a set-up script, in line with those in tools/extras/ that would get a default installation without any effort. My thinking goes, if someone "just installs" Kaldi, they want to run some training experiments, and that would rather make little sense on a small machine. I thus want to install minimum necessary dependencies. When, on the other hand, someone builds a system based on Kaldi, they already know enough to pull necessary dependencies. Am I making sense?

**Comment by danpovey** [comment]
We recently disabled 32-bit CUDA support.  I am OK with just saying 32-bit
systems are deprecated, or at least not available if you want to use MKL.

**Comment by kkm000** [comment]
Thanks. I do not think we need to note the deprecation; actually, there is a 32-bit version of MKL. It's just that it won't be installed by default, but the user can of course install in manually. But really, it's hard to think of anyone using a 32-bit platform these days.

**Comment by kkm000** [comment]
I'm doing the script to help the user install MKL on Linux (rhel, suse, debian, fedora -- painfully testing on Docker images is the trickiest part). On Windows and Macs, an installer is available (registration required but is pretty painless). Double checked, the license includes even commercial use. They plainly mention this on all download pages. I believe the installers (Mac and W.) have unattended mode to roll out on nodes; static linking is also an option, albeit quite fat, but probably makes sense for the .so flavor of Kaldi libraries, as only the matrix/ and cudamatrix/ have direct dependencies on the math lib.

One thing that sets it aside from other Kaldi-sophy scripts is it's installation location is system-wide, and is fixed to /opt/intel that likely requires a sudo privilege. I think it should be ok? It may be possible to to override the location, but that means using the system's package manager to pull raw packages (rpm or deb) and then installing them using respective rpm/dpkg tools, and I do not think we can support that in the long term. Looks like a big, big hack to me.

Speaking of documentation: is it sensible to just go ahead to encourage the use of MKL as the default matrix library choice? I would, by all means. ATLAS has been too capricious for me (never touched it for years tho, dunno if they've improved it); with OpenBLAS I have no experience. MKL is not optimized for AMD cores (ha, would it be!), but at the least uses all the instruction sets available there; I do not think it would be worse than alternatives. It would be probably illegal and certainly unethical for Intel to cripple the libraries on other CPUs, they just don't care optimizing them for other than own CPUs. Do you have any data or rough understanding of AMD use in the field with Kaldi, and what's your feeling about this set-up (AMD CPU + MKL, namely)?

**Comment by danpovey** [comment]
I have heard good things about both MKL and OpenBLAS.  What I want is to
make it easier to install Kaldi with the system-package versions of both of
those libraries (i.e. not compiling OpenBLAS ourselves, as we previous did).

I am OK with a setup script where you have to install something using, say,
apt-get as root; Kaldi already has various expectations that certain
system-level packages will exist.  We should make it easy, though, to run
the configure script when the package that it expects by default does not
exist.  E.g. have it print out suggestions, or have the check_dependencies
script print out suggestions.

Personally I would be OK with either mkl or OpenBLAS being the default, as
long as it's easy to do.  What's easiest might depend on the specific OS.

Dan

**Comment by kkm000** [comment]
Thanks, I think I got the overall idea.

**Comment by kkm000** [comment]
@danpovey, @jtrmal, quick question. The check_dependencies.sh checks for libatlas and recommends it if not found. I want to kill that check, and recommend MKL instead. So we are kind of making MKL the default choice instead of atlas. No objections?

Longer: I downloaded debian packages (used by Ubuntu & friends) for openblas, cblas and atlas; they are unuseful out of the box (the cblas.h file is renamed in all). And there is no clapack header files at all for liblapack{3,-dev}. I don't think there is an option to get Kaldi compiled out of the box w/o compile errors except MKL. E. g., Atlas has `#include "cblas.h"` in one file, but it's .h file is actually named "cblas-atlas.h". So the package is not useful at all.

See what's there for `cblas[-_a-z].h`. This is a sheer mess; no "cblas.h" anywhere.
```
kkm@yupana:~/work/mklver$ ls -1 *-dev_*
libatlas-base-dev_3.10.3-5_amd64.deb
libatlas-cpp-0.6-dev_0.6.3-4ubuntu1_amd64.deb
libblas-dev_3.7.1-4ubuntu1_amd64.deb
liblapack-dev_3.7.1-4ubuntu1_amd64.deb
libopenblas-dev_0.2.20+ds-4_amd64.deb
kkm@yupana:~/work/mklver$ echo *-dev_* | xargs -n1 dpkg-deb --contents | grep -P '/cblas[-\w]*\.h$'
-rw-r--r-- root/root     33962 2017-09-13 13:29 ./usr/include/x86_64-linux-gnu/cblas-atlas.h
-rw-r--r-- root/root     29918 2017-11-28 11:24 ./usr/include/x86_64-linux-gnu/cblas-netlib.h
-rw-r--r-- root/root     21243 2017-06-17 15:46 ./usr/include/x86_64-linux-gnu/cblas_f77.h
-rw-r--r-- root/root       444 2017-11-28 11:24 ./usr/include/x86_64-linux-gnu/cblas_mangling.h
-rw-r--r-- root/root     45648 2017-09-19 05:51 ./usr/include/x86_64-linux-gnu/cblas-openblas.h
```

**Comment by jtrmal** [comment]
I don't see any issue with that ATM.
y.

**Comment by danpovey** [comment]
I am certainly OK with a PR that tries to make MLK the default.  We would
have to have people test it out.  Please make sure it doesn't use
multi-threaded stuff by default; if it does, we need to disable it via
environment variables.  That stuff just spins on locks and wastes CPU.

The BLAS and ATLAS headers are very standardized (they don't really change)
so I would rather just use our own copies of them, it is simpler than
figuring out the packaging on every variant of linus.

**Comment by kkm000** [comment]
The biggest trouble with packaged libraries is they are compiled for the least common CPU, and for matrix algebra this is a killer. I would not go as far as carrying headers. I'm tempted to say "use MKL, or use anything else you want but then know what you are doing". This presumes the user has headers matching their libraries. I believe the current setup with HAVE_CBLAS and friends is more than covers it.

I'll write an explanation in matrixwrap.dox. It's hugely outdated anyway.

> Please make sure it doesn't use multi-threaded stuff by default

It does not, in configure.sh the single-threaded lib is always selected. I've been using it with Kaldi's configure for years, no problems in this part. And yea, I know what you're talking about. I had been getting a lot of performance surprises if not worse with MKL multithreading.

**Comment by kkm000** [comment]
@danpovey, @jtrmal, I am wondering why do we have the multithreaded math options in configure at all. Are they beneficial in any scenario at all? The way Kaldi is organized, you just run as many processes per machine as to use up its CPU power, and threading math would almost certainly not be helpful. I am assuming (maybe not entirely correctly) that people who integrate Kaldi in their own code would heavily customize their compilation anyway, not just plop the libraries. I, for one, use a lot of optimization trickery (such as profile-guided optimization) to get our code to production in the most efficient shape.

The moment I gave up even trying to use MKL multithreading was when I got a significantly(!) worse performance on a host machine than inside a VM hosted by the same host. At least with MKL, it's surprisingly unpredictable, and we have too much variety of CPUs in our racks to use it reliably.

So, the question is are they worth keeping. To me, they have a potential to confuse the user (I imagine the Kaldi user is first of all an ML/ASR expert or student, data scientist but not necessary a computer architecture geek). But I do not have a full picture, and you do. If keep, I must explain the impact as I'm updating the docs on matrix math. If you can please explain how/when they could be useful, that would be very helpful.
 
I've got our shiny new user-friendly MKL installer working on a bunch of systems ([list](https://github.com/kkm000/install_mkl/blob/master/README.md)), so the docs is the only part left to do.

**Comment by danpovey** [comment]
Yeah, I think you're right, multithreaded BLAS is not a good idea and we
don't need to configure it to use it.

**Comment by kkm000** [comment]
So I'm going to discourage that in the docs at the very least, but what is your opinion on altogether retiring the configure option `--threaded-math`?

**Comment by danpovey** [comment]
I'm OK to retire it.

**Comment by jtrmal** [comment]
Me too -- I think it's something very (pre)historic and I don't think there
ever has been a good use case for it.
y.

**Comment by kkm000** [comment]
Anyone has a Mac handy? I'd appreciate, if you do and are bored and have a bit of time to burn for a good cause, please install MKL with the installer from https://software.intel.com/mkl/choose-download (registration needed, but painless). I would like to figure out where does it get installed under Darwin. I figured out from scripts that the libraries are in `lib/`, not in `lib/intel64` but I do not know where the root of this folder could be. I do not even know if it's always the same location, or based on some system setting. Any help is appreciated!

**Comment by kkm000** [comment]
@danpovey, @jtrmal, thanks, so I'll kill it! Yay!

I'll send a separate PR, don't want to lump two things together.

**Comment by danpovey** [comment]
```
 ls /opt/intel/mkl/lib/
libmkl_avx.dylib                libmkl_core.a
 libmkl_intel_lp64.dylib         libmkl_mc.dylib
 libmkl_tbb_thread.a             libmkl_vml_mc.dylib
libmkl_avx2.dylib               libmkl_core.dylib
 libmkl_intel_thread.a           libmkl_mc3.dylib
libmkl_tbb_thread.dylib         libmkl_vml_mc2.dylib
libmkl_avx512.dylib             libmkl_intel_ilp64.a
libmkl_intel_thread.dylib       libmkl_rt.dylib
 libmkl_vml_avx.dylib            libmkl_vml_mc3.dylib
libmkl_blas95_ilp64.a           libmkl_intel_ilp64.dylib
libmkl_lapack95_ilp64.a         libmkl_sequential.a
 libmkl_vml_avx2.dylib           locale
libmkl_blas95_lp64.a            libmkl_intel_lp64.a
 libmkl_lapack95_lp64.a          libmkl_sequential.dylib
 libmkl_vml_avx512.dylib

```

**Comment by kkm000** [comment]
X-ref: Follow-up fix in #3216.
